### DISCUSSION TITLE
Tx sender recovery with rsv signatures in Hardhat local network

### DISCUSSION THREAD

**Discussion by paul-zibarov** [NomicFoundation/hardhat#3369] [Troubleshooting]
Hello! I am trying to create an on-chain tx sender validator. Smart contract receive tx data (tx.nonce, tx.gasprice, tx.gaslimit, tx.to, tx.value, tx.data), signatures (tx.v, tx.r, tx.s) and sender(tx.from) than rlp encode tx parameters and generate hash which used in solidity ecrecover(hash, r, s, v). This function returns recovered address, which should be equal to tx.from if parameters were encoded correctly and signatures were correct. This solution works on bsc main and test networks, but I want to run some tests in local network and there is some problem with signatures.

I started hardhat local network, then published ERC-20 contract and Verifier contract, then call mint tx from the first generated account. I copied this data and pass it to Verifier and receive zero address as result. 

Why I am thinking that there is a problem with signatures?

As offchain alternative I use script that does same things to tx data. I am using these packages:

```
import { ecsign, keccak256, toBuffer, pubToAddress } from 'ethereumjs-util';
import { encode } from 'rlp';
import Wallet from 'ethereumjs-wallet';
```

I copied the private key from the generated account and sign tx.hash with it and received completely other signatures.
There is no problem with hashes, because smart contract and js rlp library generated the same hash. I passed these signatures with hash to ecrecover function and finally received generated account address.

I also tried to send raw tx with offchain-generated signatures and mined tx from a completely different address.

Example data:

```
const account = {
    address: "0xf39fd6e51aad88f6f4ce6ab8827279cfffb92266",
    pk: "0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80" //locally generated pk
}

const transaction = {
    nonce: '0x2',
    gasPrice: '0xbda8863b',
    gasLimit: '0x11303',
    to: '0x5fbdb2315678afecb367f032d93f642f64180aa3',
    value: '0x00',
    data: '0x40c10f19000000000000000000000000f39fd6e51aad88f6f4ce6ab8827279cfffb922660000000000000000000000000000000000000000000000000000000000000001',
    chainId: 2
};

//raw solidity
0xf8660284bda8863b83011303945fbdb2315678afecb367f032d93f642f64180aa300b84440c10f19000000000000000000000000f39fd6e51aad88f6f4ce6ab8827279cfffb922660000000000000000000000000000000000000000000000000000000000000001
//hash solidity
0xcfd270cff5fd1333e6bbacbe27497408ec732b0e6f0eb48cc4eeecaf3394af77
//v hardhat
0
//r hardhat
0xf45fdc2d382b048a2a456ed866eb7c5ea29692928f88b86654bd6e94d9543086
//s hardhat
0x553a8973873868552d342929a4814f1aa65ea8627d6f745efb549197cb005e92

//------------------------------------------------------------------

//raw eth-utils
0xf8660284bda8863b83011303945fbdb2315678afecb367f032d93f642f64180aa300b84440c10f19000000000000000000000000f39fd6e51aad88f6f4ce6ab8827279cfffb922660000000000000000000000000000000000000000000000000000000000000001
//hash eth-utils
0xcfd270cff5fd1333e6bbacbe27497408ec732b0e6f0eb48cc4eeecaf3394af77
//v eth-utils
27
//r eth-utils
0x95266727f6886d8099ccb11bd82d8bacba5ceb5ead47d2497998be78be74ead4
//s eth-utils
0x11a5fc2d0d0e243fb07d38af7d44bf41fe2b0523531e2d91d4207a6809810298
//raw + signature eth-utils
0xf8a90284bda8863b83011303945fbdb2315678afecb367f032d93f642f64180aa300b84440c10f19000000000000000000000000f39fd6e51aad88f6f4ce6ab8827279cfffb92266000000000000000000000000000000000000000000000000000000000000000127a095266727f6886d8099ccb11bd82d8bacba5ceb5ead47d2497998be78be74ead4a011a5fc2d0d0e243fb07d38af7d44bf41fe2b0523531e2d91d4207a6809810298 
```

My hardhat config:

```
import "@nomicfoundation/hardhat-chai-matchers";
import "@nomicfoundation/hardhat-toolbox";
import '@openzeppelin/hardhat-upgrades';
import "hardhat-abi-exporter";
import dotenv from "dotenv";

dotenv.config();

const config = {
  solidity: {
    version: "0.8.10",
    settings: {
      optimizer: {
        enabled: true,
        runs: 200,
      },
    },
  },
  networks: {
    hardhat: {
      chainId: 2
    }
  }
};

export default config;
```

I would be appreciated any tips or help, cause I am stuck in these tests.

**Comment by fvictorio** [comment]
@paul-zibarov you should also include the code you are using to generate the parameters, and the solidity code that you are using to recover the address.

**Reply by paul-zibarov** [reply]
Sure!

Solidity code:
```
function generateRaw(uint256 gasPrice, uint256 gasLimit, bytes memory value, uint256 nonce, bytes memory data, address to) internal pure returns(bytes memory raw) {
        bytes[] memory encoded = new bytes[](6);
        encoded[0] = RLPEncoder.encodeUint(nonce);
        encoded[1] = RLPEncoder.encodeUint(gasPrice);
        encoded[2] = RLPEncoder.encodeUint(gasLimit);
        encoded[3] = RLPEncoder.encodeAddress(to);
        encoded[4] = RLPEncoder.encodeBytes(value);
        encoded[5] = RLPEncoder.encodeBytes(data);

        return RLPEncoder.encodeList(encoded);
    }

    function generateHash(bytes memory raw) internal pure returns(bytes32 hash) {
        return keccak256(raw);
    }

    function verifyHash(bytes32 hash, address from, uint8 v, bytes32 r, bytes32 s) internal pure returns(bool) {
        address recovered = ecrecover(hash, v, r, s);
        return recovered == from;
    }
    
    function createProposal(uint256 requestId, Transaction calldata transaction) external {
        bytes memory raw = generateRaw(transaction.gasPrice, transaction.gasLimit, transaction.value, transaction.nonce, transaction.data, transaction.to);
        bytes32 _hash = generateHash(raw);
        bool verified = verifyHash(_hash, transaction.from, transaction.v, transaction.r, transaction.s);

        if(!verified) revert SenderNotVerified();
       ...
    }
```

Alternative js code:
```
  const transaction = await providerLocal.getTransaction(txHash);

  const wallet = Wallet.fromPrivateKey(privateKey);

  // Get a public key
  const publicKey = wallet.getPublicKey();
  const address = pubToAddress(publicKey);
  const params = [transaction.nonce, transaction.gasPrice, transaction.gasLimit, transaction.to, transaction.value, transaction.data];

  // Raw transaction as Buffer
  const rawTransaction = encode(params);
  console.log("🚀 - rawTransaction", rawTransaction.toString('hex'));
  const hash = keccak256(rawTransaction);
  console.log("🚀 - hash", hash.toString('hex'));

  // Signature of the hash of the raw transaction
  const { v, r, s } = ecsign(hash, privateKey, transaction.chainId);
  console.log(v.toString(16), r.toString('hex'), s.toString('hex'))

  // RLP encoded signed transaction as Buffer
  const signedTransaction = encode([...params, v, r, s]);
  console.log(signedTransaction.toString('hex'));

```

**Reply by fvictorio** [reply]
I'm having a hard time trying to follow this code. Any chance you can create a repository with a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example)? The more minimal, the better. That is, if you can, please remove all the code you can that doesn't reproduce the problem.

**Reply by paul-zibarov** [reply]
Sure, will reply with a link to the repository soon

**Reply by paul-zibarov** [reply]
Hi! Been away for a few days, sorry! Here is a repository with an example: https://github.com/paul-zibarov/hardhat-recover

This example works in bscTestnet, but fails for hardhat local network

**Reply by paul-zibarov** [reply]
@fvictorio Hi! Are there any updates or comments?

**Reply by fvictorio** [reply]
Looking into this now, sorry for making you wait @paul-zibarov.

**Reply by paul-zibarov** [reply]
No problem, take your time! Thanks!

**Reply by fvictorio** [reply]
@paul-zibarov I'm getting the same test failure both in Hardhat Network and in the BSC testnet.

The test fails because the recovered address is the zero address. This is the value returned by `ecrecover` when the signature is incorrect. So I'm guessing there's some bug in the `Verifier` implementation. I won't be able to help you with that though.

By the way, to be able to run the example in the BSC testnet I had to make two changes in your repo:

- I removed the `loadFixture` usage, because that only works in the Hardhat Network
- I used the same address as owner, user1 and user2, because I didn't want to fund three different accounts. But from what I see in the code, this shouldn't be too relevant.

**Reply by paul-zibarov** [reply]
I didn`t mention it before, there is some strange thing. BSC Testnet transactions have v = 229/230 (2 * chainId + 35/36) and if i replace it with regular 27/28 - contract will recover from address. I dont know if there is some problem in node instance that generates this key or directly in evm which cannot parse possible valid key. You may try it:

`  tx.v = tx.v == 229 ? 27 : tx.v == 230 ? 28 : tx.v;  `

In my main repository there are different test files for testnet and localnet, so i am shure there is no problem with loadFixture, cause i can't use it in testnet.

Different addresses just for simplify an example
